qcodes/dataset/data_set.py

```diff
@@ -1024,6 +1079,84 @@ def _load_to_dataframes(self, datadict: ParameterData) -> Dict[str, "pd.DataFram
             dfs[name] = self._data_to_dataframe(subdict, index)
         return dfs
 
+    def get_data_as_xarray(self,
+                           *params: Union[str,
+                                          ParamSpec,
+                                          _BaseParameter],
+                           concat: Optional[bool] = False,
+                           start: Optional[int] = None,
+                           end: Optional[int] = None) -> \
+            Union[Mapping[Hashable, "xr.DataArray"], "xr.Dataset"]:
```

Comment:
Why Hashable here. Is this not always a string?. I would also suggest never to use Mapping or any other generic type in return position. We know this is s dict

Reply:
You'd think `Dict[str, xr.DataArray]` would be correct but really struggled to get typing to pass. I think they're using generic types for xarray based on this [PR ](https://github.com/pydata/xarray/pull/2929#issue-274455843).

Reply:
I am confused how that can matter. We create the variable that this is the type of here 
```
data_arrs: MutableMapping[Hashable, xr.DataArray] = {}
```

Reply:
I guess the problem is 
```
qcodes\dataset\data_set.py:1150:26: error: Argument 1 to "Dataset" has incompatible type "Dict[str, DataArray]"; expected "Mapping[Hashable, Any]"
```

That should imho not be solved by changing our types. I think a better solution is to simply cast before passing to that function. 

The problem is that Mapping is not covariant in the first type so even if str is considered a subtype of Hashable
Dict[str, Any] is not a subtype of Dict[Hashable, Any] for that reason I would normally not think it worth the trouble of using Hashable for the key type until https://github.com/python/mypy/issues/1114 is resolved

Reply:
Ahh that's right the problem is with xarray's `Dataset` type being incompatible with `Dict[str, DataArray]`. Thanks for explaining the core of the issue, I was getting pretty confused why a subtype of Hashable was not working. 

I'll cast before passing into the function and leave a note about mypy#1114 typing#445.